File: wasm/sei_parser.cpp
#include "sei_parser.hpp"
#include <sstream>
#include <iomanip>
#include <cstring>

SEIParser::SEIParser() : sequentialFrameCounter(0) {}

SEIParser::~SEIParser() {}

std::vector<SEIEntry> SEIParser::parse(const uint8_t* data, size_t dataSize) {
    std::vector<SEIEntry> entries;
    sequentialFrameCounter = 0;

    // Find all NAL units in the bitstream
    auto nalUnits = findNALUnits(data, dataSize);

    for (const auto& nal : nalUnits) {
        size_t offset = nal.first;
        size_t size = nal.second;

        if (size < 2) continue; // Need at least NAL header

        const uint8_t* nalData = data + offset;

        // Parse NAL header (HEVC uses 2-byte header)
        uint8_t nalUnitType = (nalData[0] >> 1) & 0x3F;

        // Check if this is an SEI NAL unit
        if (nalUnitType == HEVC_NAL_SEI_PREFIX || nalUnitType == HEVC_NAL_SEI_SUFFIX) {
            parseSEINAL(nalData, size, entries);
        }
    }

    return entries;
}

std::vector<std::pair<size_t, size_t>> SEIParser::findNALUnits(const uint8_t* data, size_t dataSize) {
    std::vector<std::pair<size_t, size_t>> nalUnits;

    size_t i = 0;
    while (i < dataSize - 3) {
        // Look for start codes: 0x00 00 01 or 0x00 00 00 01
        if (data[i] == 0x00 && data[i + 1] == 0x00) {
            size_t startCodeLen = 0;

            if (data[i + 2] == 0x01) {
                startCodeLen = 3;
            } else if (data[i + 2] == 0x00 && i + 3 < dataSize && data[i + 3] == 0x01) {
                startCodeLen = 4;
            }

            if (startCodeLen > 0) {
                size_t nalStart = i + startCodeLen;

                // Find next start code
                size_t j = nalStart + 1;
                size_t nalEnd = dataSize;

                while (j < dataSize - 3) {
                    if (data[j] == 0x00 && data[j + 1] == 0x00 &&
                        (data[j + 2] == 0x01 || (data[j + 2] == 0x00 && j + 3 < dataSize && data[j + 3] == 0x01))) {
                        nalEnd = j;
                        break;
                    }
                    j++;
                }

                nalUnits.push_back({nalStart, nalEnd - nalStart});
                i = nalEnd;
            } else {
                i++;
            }
        } else {
            i++;
        }
    }

    return nalUnits;
}

void SEIParser::parseSEINAL(const uint8_t* nalData, size_t nalSize, std::vector<SEIEntry>& entries) {
    if (nalSize < 3) return; // Need NAL header + at least 1 byte

    // Skip 2-byte NAL header
    const uint8_t* seiData = nalData + 2;
    size_t seiSize = nalSize - 2;

    size_t pos = 0;
    while (pos < seiSize) {
        // Read payload type
        uint32_t payloadType = 0;
        while (pos < seiSize && seiData[pos] == 0xFF) {
            payloadType += 255;
            pos++;
        }
        if (pos >= seiSize) break;
        payloadType += seiData[pos++];

        // Read payload size
        uint32_t payloadSize = 0;
        while (pos < seiSize && seiData[pos] == 0xFF) {
            payloadSize += 255;
            pos++;
        }
        if (pos >= seiSize) break;
        payloadSize += seiData[pos++];

        // Extract payload
        if (pos + payloadSize > seiSize) break;

        const uint8_t* payload = seiData + pos;

        // Decode RBSP
        std::vector<uint8_t> decodedPayload = rbspDecode(payload, payloadSize);

        // Parse based on type
        if (payloadType == SEI_TYPE_USER_DATA_UNREGISTERED) {
            if (decodedPayload.size() >= 16) { // Must have UUID
                uint32_t frameNum = frameNumberFromUUID(decodedPayload.data());
                SEIEntry entry = parseUserDataUnregistered(decodedPayload.data(), decodedPayload.size(), frameNum);
                entries.push_back(entry);
            }
        } else if (payloadType == SEI_TYPE_TIME_CODE) {
            SEIEntry entry = parseTimecode(decodedPayload.data(), decodedPayload.size(), sequentialFrameCounter);
            entries.push_back(entry);
            sequentialFrameCounter++;
        }

        pos += payloadSize;

        // Skip trailing bits (RBSP stop bit and alignment)
        if (pos < seiSize && seiData[pos] == 0x80) {
            pos++;
        }
    }
}

std::vector<uint8_t> SEIParser::rbspDecode(const uint8_t* data, size_t size) {
    std::vector<uint8_t> decoded;
    decoded.reserve(size);

    for (size_t i = 0; i < size; ) {
        if (i + 2 < size && data[i] == 0x00 && data[i + 1] == 0x00 && data[i + 2] == 0x03) {
            // Emulation prevention: 0x00 00 03 -> 0x00 00
            decoded.push_back(data[i++]);
            decoded.push_back(data[i++]);
            i++; // Skip 0x03
        } else {
            decoded.push_back(data[i++]);
        }
    }

    return decoded;
}

SEIEntry SEIParser::parseUserDataUnregistered(const uint8_t* payload, size_t payloadSize, uint32_t frameNumber) {
    SEIEntry entry;
    entry.type = SEI_TYPE_USER_DATA_UNREGISTERED;
    entry.frameNumber = frameNumber;

    // First 16 bytes are UUID
    if (payloadSize < 16) {
        entry.jsonPayload = "{}";
        return entry;
    }

    // Remaining bytes are user data (should be JSON)
    size_t jsonSize = payloadSize - 16;
    if (jsonSize > 0) {
        const uint8_t* jsonData = payload + 16;
        entry.jsonPayload = std::string(reinterpret_cast<const char*>(jsonData), jsonSize);
    } else {
        entry.jsonPayload = "{}";
    }

    entry.payload.assign(payload, payload + payloadSize);
    entry.rawBytes = bytesToHex(payload, payloadSize);

    return entry;
}

SEIEntry SEIParser::parseTimecode(const uint8_t* payload, size_t payloadSize, uint32_t frameNumber) {
    SEIEntry entry;
    entry.type = SEI_TYPE_TIME_CODE;
    entry.frameNumber = frameNumber;
    entry.payload.assign(payload, payload + payloadSize);
    entry.rawBytes = bytesToHex(payload, payloadSize);

    if (payloadSize == 0) {
        entry.timecodeString = "--:--:--:--";
        return entry;
    }

    // Parse timecode bits
    BitReader reader(payload, payloadSize);

    try {
        // Based on H.265 D.2.27 Time code SEI message
        uint32_t num_clock_ts = reader.readBits(2);
        uint32_t units_field_based_flag = reader.readBits(1);
        uint32_t counting_type = reader.readBits(5);
        uint32_t full_timestamp_flag = reader.readBits(1);
        uint32_t discontinuity_flag = reader.readBits(1);
        uint32_t cnt_dropped_flag = reader.readBits(1);
        uint32_t n_frames = reader.readBits(9);
        
        uint32_t seconds = 0;
        uint32_t minutes = 0;
        uint32_t hours = 0;

        if (full_timestamp_flag) {
            seconds = reader.readBits(6);
            minutes = reader.readBits(6);
            hours = reader.readBits(5);
        } else {
            // Partial timestamp parsing would go here
            // For simplicity, we'll leave at 0
        }

        // Build timecode string HH:MM:SS:FF
        std::stringstream ss;
        ss << std::setfill('0') << std::setw(2) << hours << ":"
           << std::setfill('0') << std::setw(2) << minutes << ":"
           << std::setfill('0') << std::setw(2) << seconds << ":"
           << std::setfill('0') << std::setw(2) << n_frames;
        
        entry.timecodeString = ss.str();
    } catch (...) {
        entry.timecodeString = "--:--:--:--";
    }

    return entry;
}

uint32_t SEIParser::frameNumberFromUUID(const uint8_t* uuid) {
    // UUID is constructed with frame number in first 4 bytes (big-endian)
    return (uuid[0] << 24) | (uuid[1] << 16) | (uuid[2] << 8) | uuid[3];
}

std::string SEIParser::bytesToHex(const uint8_t* data, size_t size) {
    std::stringstream ss;
    ss << std::hex << std::setfill('0');
    for (size_t i = 0; i < size && i < 32; i++) { // Limit to first 32 bytes
        ss << std::setw(2) << static_cast<int>(data[i]) << " ";
    }
    if (size > 32) {
        ss << "...";
    }
    return ss.str();
}

// BitReader implementation
SEIParser::BitReader::BitReader(const uint8_t* data, size_t size)
    : data(data), size(size), bytePos(0), bitPos(0) {}

uint32_t SEIParser::BitReader::readBits(int numBits) {
    uint32_t value = 0;
    
    for (int i = 0; i < numBits; i++) {
        if (bytePos >= size) {
            return value; // End of data
        }

        // Read bit from current position
        int bit = (data[bytePos] >> (7 - bitPos)) & 1;
        value = (value << 1) | bit;

        bitPos++;
        if (bitPos >= 8) {
            bitPos = 0;
            bytePos++;
        }
    }

    return value;
}

bool SEIParser::BitReader::hasMoreBits() const {
    return bytePos < size;
}

File: src/parsers/seiParser.js
// SEI Parser - Extracts SEI data from HEVC/H.265 Annex-B streams

export class SEIParser {
    parseSEIFromAnnexB(data) {
        const seiData = [];
        let frameNumber = 0;
        let offset = 0;

        while (offset < data.length - 4) {
            // Find start code (0x00 0x00 0x00 0x01 or 0x00 0x00 0x01)
            if (data[offset] === 0x00 && data[offset + 1] === 0x00) {
                let startCodeLen = 0;
                if (data[offset + 2] === 0x00 && data[offset + 3] === 0x01) {
                    startCodeLen = 4;
                } else if (data[offset + 2] === 0x01) {
                    startCodeLen = 3;
                } else {
                    offset++;
                    continue;
                }

                offset += startCodeLen;
                if (offset >= data.length) break;

                // Read NAL unit header (2 bytes for HEVC)
                const nalHeader = (data[offset] << 8) | data[offset + 1];
                const nalType = (nalHeader >> 9) & 0x3F;

                // HEVC NAL unit types: 39 (PREFIX_SEI), 40 (SUFFIX_SEI)
                // Also check for VCL NAL units to track frame numbers
                if (nalType >= 0 && nalType <= 31) {
                    // VCL NAL unit (coded slice)
                    frameNumber++;
                } else if (nalType === 39 || nalType === 40) {
                    // SEI NAL unit found
                    offset += 2; // Skip NAL header

                    // Find next start code to determine NAL unit length
                    let nalEnd = offset;
                    while (nalEnd < data.length - 3) {
                        if (data[nalEnd] === 0x00 && data[nalEnd + 1] === 0x00 &&
                            (data[nalEnd + 2] === 0x01 || (data[nalEnd + 2] === 0x00 && data[nalEnd + 3] === 0x01))) {
                            break;
                        }
                        nalEnd++;
                    }

                    const seiPayload = data.slice(offset, nalEnd);
                    this.parseSEIPayload(seiPayload, frameNumber, seiData);
                    offset = nalEnd;
                    continue;
                }

                offset++;
            } else {
                offset++;
            }
        }

        return seiData;
    }

    parseSEIPayload(payload, frameNumber, seiData) {
        let offset = 0;

        while (offset < payload.length) {
            // Read payload type (can be multi-byte)
            let payloadType = 0;
            while (offset < payload.length && payload[offset] === 0xFF) {
                payloadType += 255;
                offset++;
            }
            if (offset >= payload.length) break;
            payloadType += payload[offset++];

            // Read payload size (can be multi-byte)
            let payloadSize = 0;
            while (offset < payload.length && payload[offset] === 0xFF) {
                payloadSize += 255;
                offset++;
            }
            if (offset >= payload.length) break;
            payloadSize += payload[offset++];

            if (offset + payloadSize > payload.length) break;

            const payloadData = payload.slice(offset, offset + payloadSize);

            // Parse specific SEI types
            if (payloadType === 136 || payloadType === 0x88) {
                // Time code SEI
                const timecode = this.parseTimecodeSEI(payloadData);
                if (timecode) {
                    seiData.push({
                        frameNumber,
                        type: payloadType,
                        ...timecode
                    });
                }
            } else if (payloadType === 5 || payloadType === 0x05) {
                // User data unregistered SEI
                const userData = this.parseUserDataSEI(payloadData);
                if (userData) {
                    seiData.push({
                        frameNumber,
                        type: payloadType,
                        ...userData
                    });
                }
            }

            offset += payloadSize;

            // Skip trailing bits (0x80 byte alignment)
            if (offset < payload.length && payload[offset] === 0x80) {
                offset++;
            }
        }
    }

    parseTimecodeSEI(data) {
        try {
            if (data.length < 5) return null;

            // Create a bit reader
            let bitOffset = 0;
            const readBits = (numBits) => {
                let result = 0;
                for (let i = 0; i < numBits; i++) {
                    const byteIndex = Math.floor(bitOffset / 8);
                    const bitIndex = 7 - (bitOffset % 8);
                    if (byteIndex >= data.length) return 0;
                    const bit = (data[byteIndex] >> bitIndex) & 1;
                    result = (result << 1) | bit;
                    bitOffset++;
                }
                return result;
            };

            // Read according to the minimal H265 format
            // Skip a few header bits according to minimal H265 timecode layout
            readBits(2);
            readBits(1);
            readBits(1);
            readBits(5);
            readBits(1);
            readBits(1);
            readBits(1);

            // The actual timecode values
            const frames = readBits(9);   // frames_counter, size=9
            const seconds = readBits(6);  // seconds_counter, size=6
            const minutes = readBits(6);  // minutes_counter, size=6
            const hours = readBits(5);    // hours_counter, size=5
            readBits(5); // skip reserved bits

            // Build timecode string HH:MM:SS:FF
            const timecodeString = `${hours.toString().padStart(2, '0')}:${minutes.toString().padStart(2, '0')}:${seconds.toString().padStart(2, '0')}:${frames.toString().padStart(2, '0')}`;

            return {
                timecodeString,
                hours,
                minutes,
                seconds,
                frames,
                dropFrame: false
            };
        } catch (e) {
            return null;
        }
    }

    parseUserDataSEI(data) {
        try {
            // Skip UUID (16 bytes) if present
            let offset = 0;
            if (data.length > 16) {
                offset = 16;
            }

            // Try to decode remaining as UTF-8 string (often JSON)
            const textData = data.slice(offset);
            let jsonPayload = new TextDecoder('utf-8').decode(textData);
            jsonPayload = jsonPayload.replace(/[^\x20-\x7E]/g, '');
            jsonPayload = jsonPayload.trim();

            return {
                jsonPayload
            };
        } catch (e) {
            return null;
        }
    }
}
